### PR TITLE
Use SPM breaking changes checker

### PR DESCRIPTION
**Pull request by FranzBusch**
### Motivation:

SPM has built in functionality to check the API of modules against a target git treeish. We can use this to simplify our `check_no_api_breakages.sh` script. Closes https://github.com/apple/swift-nio/issues/1239

### Modifications:

This PR, exchanges the direct calls to Swift's API checker with the new SPM `diagnose-api-breaking-changes` tool. This allows us to get rid of the manual module parsing, build invocations and result comparisons.

### Result:

We are now using SPMs `diagnose-api-breaking-changes` to check for breaking changes.

### CODE REVIEW

**Comment by FranzBusch** [comment]
@swift-nio-bot please test

**Comment by FranzBusch** [comment]
@swift-server-bot test this please

**Comment by FranzBusch** [comment]
@swift-server-bot test this please

**Comment by FranzBusch** [comment]
@swift-server-bot test this please

File: scripts/check_no_api_breakages.sh
Comment (weissi) [inline scripts/check_no_api_breakages.sh:19]: this isn't ideal as it will only work if `origin` is actually the upstream remote.

Comment (FranzBusch) [reply scripts/check_no_api_breakages.sh:19]: I know this is not ideal. I really just want to use just `main` but that does not work. I guess the CI is doing a shallow clone and `main` is not fetched in that case. I am still trying some things out to find a good config that works both remotely and locally.

**Comment by FranzBusch** [comment]
Here are two example runs one that failed and one that succeeded:
https://ci.swiftserver.group/job/swift-nio2-api-breakage-prb/2579/console
https://ci.swiftserver.group/job/swift-nio2-api-breakage-prb/2580/console

**Comment by FranzBusch** [comment]
@weissi @Lukasa since the `diagnose-api-breaking-changes` is only available in the nightly toolchains I would suggest us to move the api breakage pipeline to the `docker-compose.2004.main.yaml` docker setup. I also tried to use the experimental tool that is shipped with 5.5 but I ran into several errors that have been fixed on main already. What do you think about using the other docker setup for the api breaking checks?

File: scripts/check_no_api_breakages.sh
Comment (weissi) [reply scripts/check_no_api_breakages.sh]: Maybe @tomerd can help here. I’m pretty sure that CI would somehow be able to refer to the base branch for the PR. Technically `main` is also not correct (but correct enough probably) because you want to compare to the target branch your PR goes into.

Comment (FranzBusch) [reply scripts/check_no_api_breakages.sh]: With my latest changes it actually works and I just refactored your script to call out to SPM instead. I does a separate clone still. If we just want to simply invoke SPM with `main` then we need some changes on the CI probably.

**Comment by glbrntt** [comment]
> @weissi @Lukasa since the `diagnose-api-breaking-changes` is only available in the nightly toolchains I would suggest us to move the api breakage pipeline to the `docker-compose.2004.main.yaml` docker setup. I also tried to use the experimental tool that is shipped with 5.5 but I ran into several errors that have been fixed on main already. What do you think about using the other docker setup for the api breaking checks?

I'd rather stick to a released version to avoid dealing with potential problems associated with nightly toolchains (compiler crashes etc.) and revisit this when the tool is no longer experimental.

**Comment by FranzBusch** [comment]
@glbrntt I spoke with @fabianfett yesterday about this as well and we came to the same conclusion. Let's put this PR on hold until the tool lands in a released Swift version and then we can revisit it.

File: scripts/check_no_api_breakages.sh
Comment (weissi) [inline scripts/check_no_api_breakages.sh:55]: that line is superfluous but also doesn't hurt ;)

**Review by weissi — APPROVED** [review]
This is fantastic, thank you!

**Pull request merged**
